### PR TITLE
feat: multiple inherited constructors in abstract contracts

### PR DESCRIPTION
**Pull request by agusduha**
# 🤖 Linear

Closes BES-318

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -289,12 +289,14 @@ export async function getSourceUnits(rootPath: string, contractsDirectories: str
...
 
-export function smockableNode(node: ASTNode): boolean {
+export function smockableNode(node: ASTNode, contract: ContractDefinition): boolean {
   if (node instanceof VariableDeclaration) {
     // If the state variable is constant then we don't need to mock it
     if (node.constant || node.mutability === 'immutable') return false;
     // If the state variable is private we don't mock it
     if (node.visibility === 'private') return false;
+  } else if (node instanceof FunctionDefinition) {
+    if (node.isConstructor && contract.abstract) return false;
```
Comment (gas1cent) [inline src/utils.ts]: Can we access the contract as `parent` of the node?

File: src/utils.ts [hunk 2]
```diff
@@ -314,14 +316,24 @@ export async function renderAbstractUnimplementedFunctions(contract: ContractDef
...
+
   for (const selector in inheritedSelectors) {
     // Skip the functions that are already implemented in the current contract
     if (currentSelectors.includes(selector)) continue;
 
     // Skip the functions that are already implemented in the inherited contracts
     if (inheritedSelectors[selector].implemented) continue;
 
-    const func = inheritedSelectors[selector].function;
+    const func = inheritedSelectors[selector].functions[0];
```
Comment (gas1cent) [inline src/utils.ts]: @agusduha could you clarify the changes to this code? In particular, I have 2 questions:
- Why `function` has turned into `functions`
- Why getting the first function here?

File: src/utils.ts (hunk 1 above)
Comment (agusduha) [reply src/utils.ts]: Yeah its a good improvement

File: src/utils.ts (hunk 2 above)
Comment (agusduha) [reply src/utils.ts]: The thing is that now we want to save all the constructors functions definitions, but we still don't need them for a common function. For functions that are not constructor there is a list with only 1 node. Maybe to make this better we have to separate functions from constructors, anyway we still need to have ona base node to not change the context function.

**Review by gas1cent — APPROVED** [review]
🍻

**Pull request merged**